### PR TITLE
[BEAM-3861] Improve test infra in Python SDK for streaming end-to-end test

### PR DESCRIPTION
**Pull request by markflyhigh**
This is the first part of work to enable Python streaming end-to-end test. Changes include:

- Improved test runner to make sure that test execution is unblocked after job is submitted successfully and start running. This only applies to streaming tests.
- `streaming_wordcount_it_test.py` is created as an example of end-to-end streaming test. The test runs `streaming_wordcount.py` and publish numbers to PubSub as test data. Currently, the test only verifies that job can successfully enter RUNNING state.
- Add few PubSub utils.

Notes: 
- currently, `streaming_wordcount_it_test` needs to be manually terminate, otherwise, it'll run forever.

TODO:
- Add verifier to the test to check output PubSub messages are expected.
- Auto-cancel pipeline in end-to-end test for streaming.
- Improve wait_until_running condition.

Following changes are made in another [PR](https://github.com/markflyhigh/incubator-beam/pull/2).

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by markflyhigh** [comment]
+cc: @aaltay

**Comment by markflyhigh** [comment]
Rebased against master. It's ready to review.
Please take a look @aaltay

File: sdks/python/apache_beam/examples/streaming_wordcount_it_test.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""End-to-end test for the streaming wordcount example.
+
+Important: End-to-end test infrastructure for streaming pipeine in Python SDK
+is in development and is not yet available for use.
+
+Currently, this test is blocked until manually terminate the pipeline job.
```
Comment (aaltay) [inline sdks/python/apache_beam/examples/streaming_wordcount_it_test.py]: How about:

Currently, this test blocks until the job is manually terminated.

File: sdks/python/apache_beam/examples/streaming_wordcount_it_test.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+  def _inject_numbers(self, topic, num_messages):
+    """Inject numbers as test data to PubSub."""
+    logging.debug('Injecting %d numbers to topic %s',
+                  num_messages, topic.full_name)
+    for n in range(num_messages):
+      topic.publish(str(n))
+
+  def _cleanup_pubsub(self):
+    test_utils.cleanup_subscription([self.input_sub, self.output_sub])
```
Comment (aaltay) [inline sdks/python/apache_beam/examples/streaming_wordcount_it_test.py]: Rename to `cleanup_subscriptions`?

File: sdks/python/apache_beam/examples/streaming_wordcount_it_test.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  @attr('developing_test')
+  def test_streaming_wordcount_it(self):
+    # Set extra options to the pipeline for test purpose
+    pipeline_verifiers = [PipelineStateMatcher(PipelineState.RUNNING)]
+    extra_opts = {'input_sub': self.input_sub.full_name,
+                  'output_topic': self.output_topic.full_name,
+                  'on_success_matcher': all_of(*pipeline_verifiers)}
+
+    # Generate input data and inject to PubSub.
+    test_utils.wait_for_subscriptions_created([self.input_sub])
```
Comment (aaltay) [inline sdks/python/apache_beam/examples/streaming_wordcount_it_test.py]: You already created subscriptions in `setUp`. Do you need to check it here?

File: sdks/python/apache_beam/testing/test_utils.py [hunk 4]
```diff
@@ -129,3 +131,47 @@ def delete_files(file_paths):
...
+def wait_for_topics_created(topics, timeout=60):
+  """Wait for all PubSub topics are created."""
+  return _wait_until_all_exist(topics, timeout)
+
+
+def _wait_until_all_exist(components, timeout):
+  need_wait = set(components)
+  start_time = time.time()
+  while time.time() - start_time <= timeout:
+    for c in components:
```
Comment (aaltay) [inline sdks/python/apache_beam/testing/test_utils.py]: I think you can do `for c in needs_wait` instead to iterate over a shorter list and remove `c in need_wait` below.

File: sdks/python/apache_beam/testing/test_utils.py [hunk 5]
```diff
@@ -129,3 +131,47 @@ def delete_files(file_paths):
...
+      if c in need_wait and c.exists():
+        need_wait.remove(c)
+    if len(need_wait) == 0:
+      return True
+    time.sleep(2)
+
+  raise RuntimeError('Timeout after %d seconds. %d of %d topics/subscriptions '
```
Comment (aaltay) [inline sdks/python/apache_beam/testing/test_utils.py]: Would it be helpful if you log what does not exist?

File: sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py [hunk 6]
```diff
@@ -46,10 +52,39 @@ def run_pipeline(self, pipeline):
...
     return self.result
+
+  def _is_in_terminate_state(self, job_state):
+    return job_state in [
+        PipelineState.STOPPED, PipelineState.DONE,
+        PipelineState.FAILED, PipelineState.CANCELLED,
+        PipelineState.UPDATED, PipelineState.DRAINED,
+    ]
+
+  def wait_until_running(self):
```
[the remark this answers is not in the document]
Comment (markflyhigh) [reply sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py]: That's exactly what I did in the following [PR](https://github.com/markflyhigh/incubator-beam/pull/2/files#diff-e48ff65b992d87457791341246d796f9R83). How about we generalize it in the future?

File: sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py [hunk 7]
```diff
@@ -46,10 +52,39 @@ def run_pipeline(self, pipeline):
       print (
           'Found: https://console.cloud.google.com/dataflow/jobsDetail'
           '/locations/%s/jobs/%s?project=%s' % (region_id, job_id, project))
-    self.result.wait_until_finish()
+
+    if not options.view_as(StandardOptions).streaming:
+      self.result.wait_until_finish()
+    else:
+      # TODO: Ideally, we want to wait until workers start successfully.
+      self.wait_until_running()
```
[the remark this answers is not in the document]
Comment (markflyhigh) [reply sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py]: I feel this is one preferred step of streaming test before executing verifiers. In integration test class, people can use `PipelineStateMatcher` to get a clear signal that the pipeline entered in RUNNING state (like [here](https://github.com/apache/beam/pull/4874/files#diff-a7940d9c8528322f7e20488c31ee47d3R85)). So they are confident to define/add following actions (like customized verifiers).

File: sdks/python/apache_beam/examples/streaming_wordcount_it_test.py (hunk 3 above)
Comment (markflyhigh) [reply sdks/python/apache_beam/examples/streaming_wordcount_it_test.py]: Actually I want to wait and verify the subscription exists before injecting data to PubSub in following line. From the pubsub API, there seems no guarantee that subscription is created successfully or instantly after calling `subscription.create()`. If subscription is created after data injection starts, data may be missing and verification will not be correct.

File: sdks/python/apache_beam/testing/test_utils.py (hunk 4 above)
Comment (markflyhigh) [reply sdks/python/apache_beam/testing/test_utils.py]: If I replace `components` with `needs_wait` in line 150, I'll get:

RuntimeError: Set changed size during iteration.

guess we cannot modify set during iteration.

File: sdks/python/apache_beam/examples/streaming_wordcount_it_test.py (hunk 1 above)
Comment (markflyhigh) [reply sdks/python/apache_beam/examples/streaming_wordcount_it_test.py]: done

File: sdks/python/apache_beam/examples/streaming_wordcount_it_test.py (hunk 2 above)
Comment (markflyhigh) [reply sdks/python/apache_beam/examples/streaming_wordcount_it_test.py]: done

File: sdks/python/apache_beam/testing/test_utils.py (hunk 5 above)
Comment (markflyhigh) [reply sdks/python/apache_beam/testing/test_utils.py]: sure. done.

File: sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aaltay) [reply sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py]: Sounds good.

File: sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (aaltay) [reply sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py]: I understand this, but for verification I think we need more than that. It is fine as one step.

File: sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py [hunk 8]
```diff
@@ -46,10 +52,39 @@ def run_pipeline(self, pipeline):
...
 
     if on_success_matcher:
       from hamcrest import assert_that as hc_assert_that
       hc_assert_that(self.result, pickler.loads(on_success_matcher))
 
     return self.result
+
+  def _is_in_terminate_state(self, job_state):
```
[the remark this answers is not in the document]
Comment (aaltay) [reply sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py]: It is fair to make it public. (We can also use it later to address this todo: https://github.com/apache/beam/blob/b5041e9ab5c5fbb4435edb6fdcab78bc3483e86c/sdks/python/apache_beam/runners/dataflow/dataflow_runner.py#L989)

File: sdks/python/apache_beam/testing/test_utils.py (hunk 4 above)
Comment (aaltay) [reply sdks/python/apache_beam/testing/test_utils.py]: OK. Maybe add a TODO to clean this up later.

File: sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (markflyhigh) [reply sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py]: sg, will change.

File: sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (markflyhigh) [reply sdks/python/apache_beam/runners/dataflow/test_dataflow_runner.py]: can you explain more about what other verification we can do here? Probably good to do it here.

**Pull request merged**